### PR TITLE
[GraphBolt] Add csc sampling graph python side code

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
This is the python code about csc sampling graph.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/setup.py [hunk 1]
```diff
@@ -214,7 +236,19 @@ def config_cython():
...
+                    bolt_name = os.path.basename(bolt_path)
+                    os.makedirs(
+                        os.path.join(CURRENT_DIR, "dgl", "graphbolt"),
+                        exist_ok=True,
+                    )
+                    shutil.copy(
+                        os.path.join(dir_, "graphbolt", bolt_name),
+                        os.path.join(CURRENT_DIR, "dgl", "graphbolt"),
+                    )
+                    fo.write("include dgl/graphbolt/%s\n" % bolt_name)
```
Comment (jermainewang) [inline python/setup.py:249]: Can we make L226-L238 a function to avoid copy-paste?

File: python/setup.py [hunk 2]
```diff
@@ -252,6 +286,20 @@ def config_cython():
...
+                            os.path.join(
+                                os.path.dirname(
+                                    os.path.relpath(path, CURRENT_DIR)
+                                ),
+                                "graphbolt",
+                                BOLT_LIB_PATTERN,
+                            )
+                        ),
+                    )
+                )
```
Comment (jermainewang) [inline python/setup.py:300]: Similarly, we could refactor the code to make it more reusable.

File: python/dgl/dataloading2/__init__.py [hunk 3]
```diff
@@ -1,3 +1,41 @@
...
 
-from .itemset import *
+import torch
+
+from .._ffi import libinfo
+from .data_fetcher import *
+from .graph_storage import *
+from .minibatch_sampler import *
+from .negative_sampler import *
+from .subgraph_sampler import *
```
Comment (jermainewang) [inline python/dgl/dataloading2/__init__.py:12]: Do these module exist?

File: python/dgl/dataloading2/graph_storage/base.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+"""Graph bolt sampling graph module."""
+
+
+class Graph(object):
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/base.py]: `CSCSampingGraph`?

Comment (jermainewang) [reply python/dgl/dataloading2/graph_storage/base.py]: Python3 does not need to inherit object.

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+            returns None.
+        """
+        return (
+            self.c_csc_graph.node_types()
+            if self.c_csc_graph.is_heterogeneous()
+            else None
+        )
+
+    @property
+    def edge_types(self) -> Optional[torch.Tensor]:
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:93]: What's the difference between `edge_types` and `type_per_edge`?

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        return self.c_csc_graph.is_heterogeneous()
+
+    @property
+    def node_types(self) -> Optional[torch.Tensor]:
+        """Returns the type of each node in the graph.
+
+        Returns
+        -------
+        torch.Tensor or None
+            If the graph is heterogeneous, returns a 1D tensor of size num_nodes,
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:82]: integer tensor?

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+from typing import List, Optional, Tuple
+
+import torch
+
+from .base import Graph
+
+HeteroInfo = Tuple[List[str], List[str], torch.tensor, torch.tensor]
+
+
+class CSCSamplingGraph(Graph):
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py]: Missing methods to get the string name of node/edge types. Edge types should further be string triplets.

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 8]
```diff
@@ -0,0 +1,309 @@
+"""Csc format sampling graph."""
+# pylint: disable= invalid-name
+from typing import List, Optional, Tuple
+
+import torch
+
+from .base import Graph
+
+HeteroInfo = Tuple[List[str], List[str], torch.tensor, torch.tensor]
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:8]: Not readable. Why not creating a class?

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+        """
+        return self.c_csc_graph.num_edges()
+
+    @property
+    def csc_indptr(self) -> torch.tensor:
+        """Returns the indices pointer in the Csc graph.
+
+        Returns
+        -------
+        int
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:48]: Int or tensor?

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+        """
+        return self.c_csc_graph.csc_indptr()
+
+    @property
+    def indices(self) -> torch.tensor:
+        """Returns the indices in the Csc graph.
+
+        Returns
+        -------
+        int
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:59]: int or tensor?

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py (hunk 7 above)
Comment (jermainewang) [reply python/dgl/dataloading2/graph_storage/csc_sampling_graph.py]: Docstring of non-straightforward methods should have examples in docstring.

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+
+
+def from_csc(
+    csc_indptr: torch.Tensor,
+    indices: torch.Tensor,
+    etype_sorted: bool = True,
+    num_nodes: Optional[int] = None,
+    hetero_info: Optional[HeteroInfo] = None,
+) -> CSCSamplingGraph:
+    """
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:152]: no newline before the brief description

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+    >>> print(graph)
+    CSCSamplingGraph(csc_indptr=tensor([0, 2, 5, 7]),
+                    indices=tensor([1, 3, 0, 1, 2, 0, 3]),
+                    num_nodes=3, num_edges=7, is_heterogeneous=True)
+    """
+    if num_nodes is None:
+        num_nodes = csc_indptr.shape[0] - 1
+
+    assert csc_indptr.shape[0] == num_nodes + 1
+    assert torch.max(indices) < num_nodes
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:192]: Will this check be costly when the graph is large?

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+        return CSCSamplingGraph(
+            torch.ops.graphbolt.from_csc(num_nodes, csc_indptr, indices)
+        )
+    else:
+        if etype_sorted is False:
+            ntypes, etypes, node_type_offset, type_per_edge = hetero_info
+            # sort neighbors by edge type id for each node
+            sorted_etype = []
+            sorted_indices = []
+            for s, e in zip(csc_indptr[:-1], csc_indptr[1:]):
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:203]: Will this loop be costly when the graph is large?

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+    num_nodes: Optional[int] = None,
+    hetero_info: Optional[HeteroInfo] = None,
+) -> CSCSamplingGraph:
+    """
+    Create a CSCSamplingGraph object from a COO representation.
+
+    Parameters
+    ----------
+    coo : torch.Tensor
+        COO format graph.
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:229]: All tensor arguments MUST have shape and dtype description.

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+    Returns
+    -------
+    CSCSamplingGraph
+        The created CSCSamplingGraph object.
+
+    Examples
+    --------
+    >>> coo = torch.tensor([[0, 1, 2, 2, 3, 3],
+    ...                     [1, 2, 0, 3, 1, 3]])
+    >>> ntypes, etypes = ['n1', 'n2', 'n3'], ['e1', 'e2']
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:245]: etypes should be tuples instead of strings.

Comment (jermainewang) [reply python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:245]: The current representation is ambiguous. Don't know if its `n1, e1, n2` or `n1, e1, n3` etc.

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+    >>> graph = dataloading2.from_coo(coo, hetero_info)
+    >>> print(graph)
+    CSCSamplingGraph(csc_indptr=tensor([0, 1, 3, 4, 6]),
+                    indices=tensor([2, 0, 3, 1, 2, 3]),
+                    num_nodes=4, num_edges=6, is_heterogeneous=True)
+    """
+    assert coo.dim() == 2
+
+    if num_nodes is None:
+        num_nodes = torch.max(coo) + (1 if coo.size(1) > 0 else 0)
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:256]: Again, be careful for all these checks. Remember this class is designed for large graphs.

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+    rol = coo[1]
+    _, indices = torch.sort(rol)
+
+    row, col = torch.index_select(coo, 1, indices)
+    csc_indptr = torch.cumsum(torch.bincount(col, minlength=num_nodes), dim=0)
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:261]: Why not use DGL's existing COO to CSR operator, which is parallelized using multi-threads.

File: tests/python/pytorch/dataloading2/test_csc_sampling_graph.py [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+import backend as F
+
+import numpy as np
+import pytest
+import torch
+
+from dgl.dataloading2 import *
+
+
+def random_heterogeneous_graph_from_csc(num_nodes, num_ntypes, num_etypes):
```
Comment (jermainewang) [inline tests/python/pytorch/dataloading2/test_csc_sampling_graph.py:12]: Please fix seeds for all random generators in test. Otherwise, later contributors may see random failures.

File: tests/python/pytorch/dataloading2/test_csc_sampling_graph.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+    )
+    return torch.index_select(coo, 1, indices)
+
+
+@unittest.skipIf(
+    F._default_context_str == "gpu",
+    reason="Graph is CPU only at present.",
+)
+@pytest.mark.parametrize("num_nodes", [10, 50, 1000, 10000])
+def test_from_csc(num_nodes, num_ntypes=3, num_etypes=5):
```
Comment (jermainewang) [inline tests/python/pytorch/dataloading2/test_csc_sampling_graph.py:69]: Can you test more corner cases?

* num_ntypes=num_etypes=0 (homogenous)
* num_ntypes=1, num_etypes=many (KG)
* edge types sharing the same name, e.g., ('user', 'like', 'movie'), ('user', 'like', book'), etc.
* types without any nodes or edges
* ...

You should think hard on making your code robust.

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 20]
```diff
@@ -0,0 +1,308 @@
...
+        )
+
+    @property
+    def node_type_offset(self) -> Optional[torch.Tensor]:
+        """Returns the node type offset tensor for a heterogeneous graph.
+
+        Returns
+        -------
+        torch.Tensor or None
+            If the graph is heterogeneous, returns a 1D tensor of size num_node_types + 1,
```
Comment (frozenbugs) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:116]: line too long, max 80 chars.

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 21]
```diff
@@ -0,0 +1,308 @@
...
+            returns None.
+        """
+        return (
+            self.c_csc_graph.type_per_edge()
+            if self.c_csc_graph.is_heterogeneous()
+            else None
+        )
+
+
+def from_csc(
```
Comment (frozenbugs) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:145]: Do we expect customers to use this API directly to create this graph?

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+            hetero_info = (ntypes, etypes, node_type_offset, type_per_edge)
+
+        return CSCSamplingGraph(
+            torch.ops.graphbolt.from_csc_with_hetero_info(
+                num_nodes, csc_indptr, indices, *hetero_info
+            )
+        )
+
+
+def from_coo(
```
Comment (frozenbugs) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:218]: From API usability perspective:
Is this from_coo better or from_coo with map from `canonical edge type` to `coo torch.Tensor` better?

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 23]
```diff
@@ -0,0 +1,308 @@
...
+    --------
+    >>> coo = torch.tensor([[0, 1, 2, 2, 3, 3],
+    ...                     [1, 2, 0, 3, 1, 3]])
+    >>> ntypes, etypes = ['n1', 'n2', 'n3'], ['e1', 'e2']
+    >>> hetero_info = (ntypes, etypes, torch.tensor([0, 2, 5]), torch.tensor([0, 1, 0, 1, 0, 1]))
+    >>> graph = dataloading2.from_coo(coo, hetero_info)
+    >>> print(graph)
+    CSCSamplingGraph(csc_indptr=tensor([0, 1, 3, 4, 6]),
+                    indices=tensor([2, 0, 3, 1, 2, 3]),
+                    num_nodes=4, num_edges=6, is_heterogeneous=True)
```
Comment (frozenbugs) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:251]:
```suggestion
    CSCSamplingGraph(csc_indptr=tensor([0, 1, 3, 4, 6]),
                     indices=tensor([2, 0, 3, 1, 2, 3]),
                     num_nodes=4, num_edges=6, is_heterogeneous=True)
```

Comment (frozenbugs) [reply python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:251]: If this was the output of _csc_sampling_graph_str, also update _csc_sampling_graph_str.

File: python/setup.py
Comment (frozenbugs) [inline python/setup.py:1]: create separate PR

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 24]
```diff
@@ -0,0 +1,308 @@
...
+HeteroInfo = Tuple[List[str], List[str], torch.tensor, torch.tensor]
+
+
+class CSCSamplingGraph:
+    r"""Class for Csc sampling graph."""
+
+    def __repr__(self):
+        return _csc_sampling_graph_str(self)
+
+    def __init__(self, c_csc_graph: torch.ScriptObject):
```
Comment (Rhett-Ying) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:17]: how to create an empty graph?

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 25]
```diff
@@ -0,0 +1,308 @@
...
+
+        Returns
+        -------
+        int
+            The number of edges in the graph
+        """
+        return self.c_csc_graph.num_edges()
+
+    @property
+    def csc_indptr(self) -> torch.tensor:
```
Comment (Rhett-Ying) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:43]: why not `def indptr()` as it's always csc sampling graph?

File: tests/python/pytorch/dataloading2/test_csc_sampling_graph.py (hunk 18 above)
Comment (Rhett-Ying) [inline tests/python/pytorch/dataloading2/test_csc_sampling_graph.py:12]: why no test for homogeneous graph? I tried `num_types=1, num_etypes=1` which generates heterogeneous graph.

File: python/setup.py (hunk 1 above)
Comment (peizhou001) [reply python/setup.py:249]: sure, this file relate has been moved to https://github.com/dmlc/dgl/pull/5707

File: python/setup.py (hunk 2 above)
Comment (peizhou001) [reply python/setup.py:300]: Ditto

File: python/dgl/dataloading2/__init__.py
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/dataloading2/__init__.py:1]: Move these changes to PR https://github.com/dmlc/dgl/pull/5711

File: python/dgl/dataloading2/__init__.py (hunk 3 above)
Comment (peizhou001) [reply python/dgl/dataloading2/__init__.py:12]: They are empty now, remove these from this PR.

File: python/dgl/dataloading2/graph_storage/csc_sampling_graph.py [hunk 26]
```diff
@@ -0,0 +1,313 @@
...
+
+def from_csc(
+    csc_indptr: torch.Tensor,
+    indices: torch.Tensor,
+    etype_sorted: bool = True,
+    num_nodes: Optional[int] = None,
+    hetero_info: Optional[HeteroInfo] = None,
+) -> CSCSamplingGraph:
+    """
+    Create a CSCSamplingGraph object from a CSC representation.
```
Comment (jermainewang) [inline python/dgl/dataloading2/graph_storage/csc_sampling_graph.py:154]: How to specify whether the created graph is stored in int32 or int64?

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 27]
```diff
@@ -0,0 +1,313 @@
...
+            If the graph is homogeneous, returns None.
+        """
+        return (
+            self.c_csc_graph.node_types()
+            if self.c_csc_graph.is_heterogeneous()
+            else None
+        )
+
+    @property
+    def edge_types(self) -> Optional[List[str]]:
```
Comment (jermainewang) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:93]: Is this return type still correct?

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 28]
```diff
@@ -0,0 +1,267 @@
...
+
+class GraphMetadata:
+    r"""Class for metadata of csc sampling graph."""
+
+    def __init__(
+        self,
+        node_type_to_id: Dict[str, int],
+        edge_type_to_id: Dict[Tuple[str, str, str], int],
+    ):
+        """
```
Comment (jermainewang) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:16]: no newline

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 29]
```diff
@@ -0,0 +1,267 @@
...
+    r"""Class for CSC sampling graph."""
+
+    def __repr__(self):
+        return _csc_sampling_graph_str(self)
+
+    def __init__(
+        self, c_csc_graph: torch.ScriptObject, metadata: Optional[GraphMetadata]
+    ):
+        self.c_csc_graph = c_csc_graph
+        self.metadata = metadata
```
Comment (jermainewang) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:78]: I wonder when will `metadata` be None? Technically even for homogeneous graph, there is one node type and one edge type.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 30]
```diff
@@ -0,0 +1,267 @@
...
+
+        Notes
+        -------
+        It is assumed that edges of each node are already sorted by edge type
+        ids.
+        """
+        return self.c_csc_graph.indices()
+
+    @property
+    def node_type_to_id(self) -> Optional[Dict[str, int]]:
```
Comment (jermainewang) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:132]: Could consider just return the metadata object to users to save the effort of adding those attributes. Of course, you should document important attributes in the metadata class.

**Review by jermainewang — APPROVED** [review]
Mostly LGTM. Some minor comments.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 31]
```diff
@@ -0,0 +1,267 @@
+"""CSC format sampling graph."""
+# pylint: disable= invalid-name
+from typing import Dict, Optional, Tuple
+
+import torch
+
+
+class GraphMetadata:
```
Comment (jermainewang) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:8]: Could consider refactor this to a pydantic object for saving/loading to/from JSON later. @Rhett-Ying

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 32]
```diff
@@ -0,0 +1,267 @@
...
+        node_type_to_id: Dict[str, int],
+        edge_type_to_id: Dict[Tuple[str, str, str], int],
+    ):
+        """
+        Initialize the GraphMetadata object.
+
+        Parameters
+        ----------
+        node_type_to_id : Dict[str, int]
+            Dictionary from node types to node IDs.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:22]: node id? node type id?

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 33]
```diff
@@ -0,0 +1,267 @@
...
+        edge_type_to_id : Dict[Tuple[str, str, str], int]
+            Dictionary from edge types to edge IDs.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:24]: edge id or edge type id?

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 34]
```diff
@@ -0,0 +1,267 @@
...
+            src, edge, dst = edge_type
+            assert isinstance(edge, str), "Edge type name should be string."
+            assert edge not in edges, f"Edge type {edge} is defined repeatedly."
+            edges.add(edge)
+            assert (
+                src in node_types
+            ), f"Unrecognized node type {src} in edge type {edge_type}"
+            assert (
+                dst in node_types
+            ), f"Unrecognized node type {dst} in edge type {edge_type}"
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:63]:
```suggestion
        # Validate node_type_to_id.
        assert all(
            isinstance(x, str) for x in node_types
        ), "Node type name should be string."
        assert all(
            isinstance(x, int) for x in node_type_ids
        ), "Node type id should be int."
        assert len(node_type_ids) == len(
            set(node_type_ids)
        ), "Multiple node types shoud not be mapped to a same id."

        # Validate edge_type_to_id.
        edges = set()
        for edge_type in edge_types:
            src, edge, dst = edge_type
            assert isinstance(edge, str), "Edge type name should be string."
            assert edge not in edges, f"Edge type {edge} is defined repeatedly."
            edges.add(edge)
            assert (
                src in node_types
            ), f"Unrecognized node type {src} in edge type {edge_type}"
            assert (
                dst in node_types
            ), f"Unrecognized node type {dst} in edge type {edge_type}"
        assert all(
            isinstance(x, int) for x in edge_type_ids
        ), "Edge type id should be int."
        assert len(edge_type_ids) == len(
            set(edge_type_ids)
        ), "Multiple edge types shoud not be mapped to a same id."
```

Comment (frozenbugs) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:63]: Reorder to keep consistency.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 29 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:78]: let's keep it optional for homogeneous graph, usually people don't think about edge type / node type when they play with homogeneous graph, make it required will lead to some confusion.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 30 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:132]: +1 just return metadata.

File: tests/python/pytorch/graphbolt/test_csc_sampling_graph.py [hunk 35]
```diff
@@ -0,0 +1,218 @@
...
+    csc_indptr[0] = 0
+    csc_indptr[-1] = num_edges
+    indices = torch.randint(0, num_nodes, (num_edges,))
+    return csc_indptr, indices
+
+
+def random_hetero_graph(num_nodes, num_edges, num_ntypes, num_etypes):
+    csc_indptr, indices = random_homo_graph(num_nodes, num_edges)
+    metadata = get_metadata(num_ntypes, num_etypes)
+    # random get node type split point
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/test_csc_sampling_graph.py:121]:
```suggestion
    # Randomly get node type split point.
```

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 36]
```diff
@@ -0,0 +1,257 @@
...
+class CSCSamplingGraph:
+    r"""Class for CSC sampling graph."""
+
+    def __repr__(self):
+        return _csc_sampling_graph_str(self)
+
+    def __init__(
+        self, c_csc_graph: torch.ScriptObject, metadata: Optional[GraphMetadata]
+    ):
+        self.c_csc_graph = c_csc_graph
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:80]: why add `_` for `metadata` while no `_` for `c_csc_graph`?

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 37]
```diff
@@ -0,0 +1,257 @@
...
+
+class GraphMetadata:
+    r"""Class for metadata of csc sampling graph."""
+
+    def __init__(
+        self,
+        node_type_to_id: Dict[str, int],
+        edge_type_to_id: Dict[Tuple[str, str, str], int],
+    ):
+
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:16]: nit: redundant line.
